[docs][pythonic config] Update IO management page (#13673)
dagster-io/dagster: docs/content/concepts/io-management/io-managers.mdx

## Relevant APIs

| Name                                                                          | Description                                                                                                                                                                                                 |
| ----------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| <PyObject module="dagster" object="ConfigurableIOManager" decorator />        | A base class used to define configurable IO managers which are also resources.                                                                                                                              |
| <PyObject module="dagster" object="ConfigurableIOManagerFactory" decorator /> | A base class used to specify configuration for more advanced I/O managers.                                                                                                                                  |

> | <PyObject module="dagster" object="IOManager" />                              | Base class for user-provided IO managers.                                                                                                                                                                   |

Reviewer: i know this is what it said before, but if i were new to IO managers i think this description "Base class for user-provided IOmanagers" meant any io manager i wrote should be off of this class

> | <PyObject object="build_input_context"/>                                      | Function for directly constructing a <PyObject object="InputContext"/>, to be passed to the <PyObject object="IOManager" method="load_input"/> method. This is designed primarily for testing purposes.     |

Reviewer 2: ```suggestion
| <PyObject object="build_input_context"/>                                      | Function for directly constructing an <PyObject object="InputContext"/>, to be passed to the <PyObject object="IOManager" method="load_input"/> method. This is designed primarily for testing purposes.     |
```

> | <PyObject object="build_output_context"/>                                     | Function for directly constructing a <PyObject object="OutputContext"/>, to be passed to the <PyObject object="IOManager" method="handle_output"/> method. This is designed primarily for testing purposes. |

Reviewer 2: ```suggestion
| <PyObject object="build_output_context"/>                                     | Function for directly constructing an <PyObject object="OutputContext"/>, to be passed to the <PyObject object="IOManager" method="handle_output"/> method. This is designed primarily for testing purposes. |
```

Dagster provides out-of-the-box IO managers for popular storage systems: AWS S3 (<PyObject module="dagster_aws.s3" object="s3_pickle_io_manager" />), Azure Blob Storage (<PyObject module="dagster_azure.adls2" object="adls2_pickle_io_manager" />), GCS (<PyObject module="dagster_gcp.gcs" object="gcs_pickle_io_manager" />), and Snowflake (<PyObject module="dagster_snowflake" object="build_snowflake_io_manager" />) - or you can write your own: either from scratch or by extending the `UPathIOManager` if you want to store data in an `fsspec`-supported filesystem.

### IO managers are provided by resources

> IO managers are provided through the [resources](/concepts/resources) system which means users can supply different IO managers for the same assets or ops in different situations. For example, you might use an in-memory IO manager for unit-testing and an S3 IO manager in production.

Reviewer 2: ```suggestion
IO managers are provided through the [resources](/concepts/resources) system which means you can supply different IO managers for the same assets or ops in different situations. For example, you might use an in-memory IO manager for unit-testing and an Amazon S3 IO manager in production.
```

Switching `user` to `you` to be consistent with the rest of this page

    return upstream_asset + [4]

> s3_resource = S3Resource()

Reviewer 3: since it's just one line, should we move this to inline?

Author: yes 👍

defs = Definitions(
    assets=[upstream_asset, downstream_asset],
    resources={
        "io_manager": ConfigurablePickledObjectS3IOManager(
            s3_resource=S3Resource(), s3_bucket="my-bucket"
        ),
    },
)
```

> This example also constructs a <PyObject module="dagster_aws.s3" object="S3Resource" /> because the <PyObject module="dagster_aws.s3" object="ConfigurablePickledObjectS3IOManager" /> depends on the S3 resource.

Reviewer 3: hmm was thinking maybe worth linking to resource-resource dependency doc section for more details (https://docs.dagster.io/master/concepts/resources#resources-which-depend-on-other-resources) but then was debating myself because we no longer say "io managers are resources" so linking might be actually more confusing 🤔

Reviewer 4: Agree on this, which is why IMO we should stick with the "IOManagers are resources" language.

Author: Added a reference to the resource-resource dep docs

To set an IO manager for a particular input, use the `input_manager_key` argument on <PyObject module="dagster" object="AssetIn" />.

> In this example,`first_asset` and `second_asset` will be stored using the default IO manager, but will be loaded as inputs to `third_asset` using the logic defined in the `PandasSeriesIOManager` (in this case loading as Pandas Series rather than python lists).

Reviewer 2: ```suggestion
In this example,`first_asset` and `second_asset` will be stored using the default IO manager, but will be loaded as inputs to `third_asset` using the logic defined in the `PandasSeriesIOManager` (in this case loading as Pandas Series rather than Python lists).
```

If you have specific requirements for where and how your outputs should be stored and retrieved, you can define your own IO manager. This boils down to implementing two functions: one that stores outputs and one that loads inputs.

> To define an IO manager, you will want to extend the <PyObject module="dagster" object="IOManager" /> class. Often times, you will want to extend the <PyObject module="dagster" object="ConfigurableIOManager"/> class (which subclasses `IOManager`) to attach a config schema to your IO manager.

Reviewer 2: ```suggestion
To define an IO manager, extend the <PyObject module="dagster" object="IOManager" /> class. Oftentimes, you will want to extend the <PyObject module="dagster" object="ConfigurableIOManager"/> class (which subclasses `IOManager`) to attach a config schema to your IO manager.
```

Here, we define a simple IO manager that reads and writes CSV values to the filesystem. It takes an optional prefix path through config.

```python file=/concepts/io_management/custom_io_manager.py startafter=start_io_manager_marker endbefore=end_io_manager_marker
from dagster import ConfigurableIOManager

class MyIOManager(ConfigurableIOManager):
    path_prefix: List[str] = []

    def _get_path(self, context) -> str:
        return "/".join(self.path_prefix + context.asset_key.path)

    def handle_output(self, context, obj):
        write_csv(self._get_path(context), obj)

    def load_input(self, context):

>         return read_csv(self._get_path(context))

Reviewer 3: can we type the context here?

```

The provided `context` argument for `handle_output` is an <PyObject module="dagster" object="OutputContext" />. The provided `context` argument for `load_input` is an <PyObject module="dagster" object="InputContext" />. The linked API documentation lists all the fields that are available on these objects.

### Using an IO manager factory

> If your IO manager is more complex, or needs to manage internal state, it may make sense to split out the IO manager definition from its configuration. In this case, you can use `ConfigurableIOManagerFactory`, which specifies config schema and implements a factory function that takes the config and returns an IO manager.

Reviewer 3: > you can use `ConfigurableIOManagerFactory`

do we have apidoc for `ConfigurableIOManagerFactory`? maybe use PyObject instead of plain code syntax here so users can click to see details

Reviewer 2: +1 ^

We should use this whenever possible. Better experience for users, but it'll also make it easy for us to keep things updated/correct.

```python file=/concepts/io_management/custom_io_manager.py startafter=start_io_manager_factory_marker endbefore=end_io_manager_factory_marker
from dagster import IOManager, ConfigurableIOManagerFactory
import requests

class ExternalIOManager(IOManager):
    def __init__(self, api_token):
        self._api_token = api_token

    def handle_output(self, context, obj):
        ...

    def load_input(self, context):
        ...

class ConfigurableExternalIOManager(ConfigurableIOManagerFactory):
    username: str
    password: str

    def create_io_manager(self, context) -> ExternalIOManager:
        api_token = requests.get(
            "https://my-api.com/token", auth=(self.username, self.password)
        ).json()
        return ExternalIOManager(api_token)

> ```

Reviewer 3: i found this example a bit abstract and hard to grasp why i'd bother with the factor pattern. maybe worth switching to a more realistic io manager example, or keep it but also link to concrete examples like "Custom filesystem-based IO Manager" down below.

Reviewer 4: Could be as simple as adding some sort of cache as private variable to the `ExternalIOManager` and note that one could not add that to a `ConfigurableIOManager` without erroring.

Author: Updated example to use a cache.

The io managers defined above will work with partitioned assets on any filesystem:

```python file=/concepts/io_management/filesystem_io_manager.py startafter=start_def_marker endbefore=end_marker
from typing import Optional

from dagster import ConfigurableIOManagerFactory, EnvVar

> class LocalPandasParquetIOManager(ConfigurableIOManagerFactory):

Reviewer 3: do we plan to explain `ConfigurableIOManagerFactory`?
